### PR TITLE
Removed parameter.h

### PR DESCRIPTION
**Pull request by gf712**
* removed List and Map implementations
* disabled model selection for now

### CODE REVIEW

File: src/shogun/io/Serializable.h [hunk 1]
```diff
@@ -145,10 +146,15 @@ template<class T> class StringListSerializable: public SGObject
...
-				"Serialized value");
+		if constexpr (type_internal::is_sg_primitive<T>::value)
+			set_generic<T>();
+		// FIXME(?): std::vector<bool> is not a container
+		// doesn't play well with the parameter framework
+		if constexpr (!std::is_same<T, bool>::value)
+			watch_param("value", &m_value);
+		else
+			SG_WARNING("std::vector<bool> is not currently serializable. "
+					   "Use SGVector<bool> instead.");
```
Comment (gf712) [inline src/shogun/io/Serializable.h:157]: @vigsterkr I copied this from the develop branch but tests still fail for SGStringList in SGObjectAll, so I am assuming there is more to this?

**Comment by gf712** [comment]
@karlnapf the gp is stuff is working but the API is quite ugly now. But at least the unittests pass!

File: src/shogun/kernel/GaussianARDKernel.cpp [hunk 2]
```diff
@@ -239,38 +238,35 @@ SGVector<float64_t> GaussianARDKernel::get_parameter_gradient_diagonal(
...
 	{
-		SG_ERROR("Can't compute derivative wrt %s parameter\n", param->m_name);
+		SG_ERROR("Can't compute derivative wrt %s parameter\n", param.first.c_str());
 		return 0.0;
 	}
 }
 
 SGMatrix<float64_t> GaussianARDKernel::get_parameter_gradient(
-		const TParameter* param, index_t index)
+		const std::pair<std::string, std::shared_ptr<const AnyParameter>>& param, index_t index)
```
Comment (karlnapf) [inline src/shogun/kernel/GaussianARDKernel.cpp:264]: nice!

**Comment by karlnapf** [comment]
> @karlnapf the gp is stuff is working but the API is quite ugly now. But at least the unittests pass!

this is fantastic! GPs unit test coverage is extensive so we can be confident there.
There are model selection unit tests ..I assume you disabled them? What about the gradient unit tests? They do work (i.e. your refactoring of the gradient extraction does the job)?

**Comment by gf712** [comment]
> this is fantastic! GPs unit test coverage is extensive so we can be confident there.
> There are model selection unit tests ..I assume you disabled them? What about the gradient unit tests? They do work (i.e. your refactoring of the gradient extraction does the job)?

model selection unit tests are disabled, the rest is working-ish. There are a couple of issues, but should be all good soon.

 @vigsterkr The problem i have now is with swig. It is generating weird code, with std::shogun::shared_ptr instead of std::shared_ptr...

File: src/shogun/io/serialization/BitseryVisitor.h [hunk 3]
```diff
@@ -27,6 +27,13 @@ namespace shogun
 				{
 					m_s.boolValue(*v);
 				}
+
+				void on(std::vector<bool>::reference* v) override
+				{
+					bool tmp = *v;
+					m_s.boolValue(tmp);
```
Comment (lisitsyn) [inline src/shogun/io/serialization/BitseryVisitor.h:34]: Oh vector<bool> is so good :)

Comment (gf712) [reply src/shogun/io/serialization/BitseryVisitor.h:34]: yea.. there should be std::vector<bool> without optimisation that behaves like the other vectors

**Comment by gf712** [comment]
@vigsterkr so close!

File: src/interfaces/swig/shogun.i [hunk 4]
```diff
@@ -50,10 +51,10 @@
 #endif
 #endif
 
+%include "std_vector.i"
 %include "shogun_ignores.i"
 %include "RandomMixin.i"
 %include "std_shared_ptr.i" 
-%shared_ptr(shogun::SparseMatrixOperator<float64_t>)
```
Comment (gf712) [inline src/interfaces/swig/shogun.i]: @vigsterkr you needed this because the shared_ptr in Library.i was misspelled (shagun instead of shogun)

**Comment by gf712** [comment]
@vigsterkr if you cherry pick, the last two commits fix swig

**Comment by gf712** [comment]
@vigsterkr before merging shared_ptr to develop one of us should do a round of `clang-tidy -checks="performance-unnecessary-value-param"` to put a const& on all those shared_ptr arguments, otherwise there will be potentially a hit on performance relative to raw pointers

**Pull request closed without merging**